gcs_contents_manager.py

```diff
@@ -428,3 +449,262 @@ def rename_file(self, old_path, new_path):
+      raise err
+    except Exception as ex:
+      raise HTTPError(500, 'Internal server error: [{}] {}'.format(type(ex), str(ex)))
+
+  def file_exists(self, path):
+    try:
+      cm, relative_path, path_prefix = self._content_manager(path)
+      if cm:
+        return cm.file_exists(relative_path)
+      return False
+    except HTTPError as err:
+      raise err
+    except Exception as ex:
+      raise HTTPError(500, 'Internal server error: [{}] {}'.format(type(ex), str(ex)))
+
+  def dir_exists(self, path):
+    if path in ['', '/']:
+      return True
+    try:
+      cm, relative_path, path_prefix = self._content_manager(path)
+      if cm:
+        return cm.dir_exists(relative_path)
+      return False
+    except HTTPError as err:
+      raise err
+    except Exception as ex:
+      raise HTTPError(500, 'Internal server error: [{}] {}'.format(type(ex), str(ex)))
+
+  def _make_model_relative(self, model, path_prefix):
+    if 'path' in model:
+      model['path'] = '{}/{}'.format(path_prefix, model['path'])
+    if model.get('type', None) == 'directory':
+      self._make_children_relative(model, path_prefix)
+
+  def _make_children_relative(self, model, path_prefix):
+      children = model.get('content', None)
+      if children:
+        for child in children:
+          self._make_model_relative(child, path_prefix)
+
+  def get(self, path, content=True, type=None, format=None):
+    if path in ['', '/']:
+      dir_obj = {}
+      dir_obj['path'] = ''
+      dir_obj['name'] = ''
+      dir_obj['type'] = 'directory'
+      dir_obj['mimetype'] = None
+      dir_obj['writable'] = False
+      dir_obj['format'] = None
+      dir_obj['content'] = None
+      dir_obj['format'] = 'json'
+      contents = []
+      for path_prefix in self._content_managers:
+        child_obj = self._content_managers[path_prefix].get('', content=False)
+        child_obj['path'] = path_prefix
+        child_obj['name'] = path_prefix
+        child_obj['writable'] = False
+        contents.append(child_obj)
+      dir_obj['content'] = contents
+      dir_obj['created'] = contents[0]['created']
+      dir_obj['last_modified'] = contents[0]['last_modified']
+      return dir_obj
+    try:
+      cm, relative_path, path_prefix = self._content_manager(path)
+      if not cm:
+        raise HTTPError(404, 'Not Found')
+
+      model = cm.get(relative_path, content=content, type=type, format=format)
+      self._make_model_relative(model, path_prefix)
+      return model
+    except HTTPError as err:
+      raise err
+    except Exception as ex:
+      raise HTTPError(500, 'Internal server error: [{}] {}'.format(type(ex), str(ex)))
+
+  def save(self, model, path):
+    if path in ['', '/']:
+      raise HTTPError(403, 'The top-level directory is read-only')
+    try:
+      self.run_pre_save_hook(model=model, path=path)
+
+      cm, relative_path, path_prefix = self._content_manager(path)
+      if not cm:
+        raise HTTPError(404, 'Not Found')
```

Comment:
Nit: similarly, can this be pushed into `_content_manager`? As far as I can tell, only `file_exists` and `dir_exists` don't need to raise 404

Reply:
Done